lib/src/sql/value/value.rs

```diff
@@ -1199,29 +1234,470 @@ impl Value {
+			Value::Number(Number::Float(ref v)) => match BigDecimal::from_f64(*v) {
+				// The Float can be represented as a Decimal
+				Some(v) => Ok(Number::Decimal(v)),
+				// Ths Float does not convert to a Decimal
+				None => Err(Error::ConvertTo {
+					from: self,
+					into: "decimal".into(),
+				}),
+			},
+			// Anything else raises an error
+			_ => Err(Error::ConvertTo {
+				from: self,
+				into: "decimal".into(),
+			}),
+		}
+	}
+
+	/// Try to coerce this value to a `Number`
+	pub(crate) fn coerce_to_number(self) -> Result<Number, Error> {
+		match self {
+			// Allow any number
+			Value::Number(v) => Ok(v),
+			// Anything else raises an error
+			_ => Err(Error::ConvertTo {
+				from: self,
+				into: "number".into(),
+			}),
+		}
+	}
+
+	/// Try to coerce this value to a `String`
+	pub(crate) fn coerce_to_string(self) -> Result<String, Error> {
+		match self {
+			// Allow any uuid value
+			Value::Uuid(v) => Ok(v.to_raw()),
+			// Allow any datetime value
+			Value::Datetime(v) => Ok(v.to_raw()),
+			// Allow any string value
+			Value::Strand(v) => Ok(v.as_string()),
+			// Anything else raises an error
+			_ => Err(Error::ConvertTo {
+				from: self,
+				into: "string".into(),
+			}),
+		}
+	}
+
+	/// Try to coerce this value to a `Strand`
+	pub(crate) fn coerce_to_strand(self) -> Result<Strand, Error> {
+		match self {
+			// Allow any uuid value
+			Value::Uuid(v) => Ok(v.to_raw().into()),
+			// Allow any datetime value
+			Value::Datetime(v) => Ok(v.to_raw().into()),
+			// Allow any string value
+			Value::Strand(v) => Ok(v),
+			// Anything else raises an error
+			_ => Err(Error::ConvertTo {
+				from: self,
+				into: "string".into(),
+			}),
+		}
+	}
+
+	/// Try to coerce this value to a `Uuid`
+	pub(crate) fn coerce_to_uuid(self) -> Result<Uuid, Error> {
+		match self {
+			// Uuids are allowed
+			Value::Uuid(v) => Ok(v),
+			// Anything else raises an error
+			_ => Err(Error::ConvertTo {
+				from: self,
+				into: "uuid".into(),
+			}),
+		}
+	}
+
+	/// Try to coerce this value to a `Datetime`
+	pub(crate) fn coerce_to_datetime(self) -> Result<Datetime, Error> {
+		match self {
+			// Datetimes are allowed
+			Value::Datetime(v) => Ok(v),
+			// Anything else raises an error
+			_ => Err(Error::ConvertTo {
+				from: self,
+				into: "datetime".into(),
+			}),
+		}
+	}
+
+	/// Try to coerce this value to a `Duration`
+	pub(crate) fn coerce_to_duration(self) -> Result<Duration, Error> {
+		match self {
+			// Durations are allowed
+			Value::Duration(v) => Ok(v),
+			// Anything else raises an error
+			_ => Err(Error::ConvertTo {
+				from: self,
+				into: "duration".into(),
+			}),
+		}
+	}
+
+	/// Try to coerce this value to a `Bytes`
+	pub(crate) fn coerce_to_bytes(self) -> Result<Bytes, Error> {
+		match self {
+			// Bytes are allowed
+			Value::Bytes(v) => Ok(v),
+			// Anything else raises an error
+			_ => Err(Error::ConvertTo {
+				from: self,
+				into: "bytes".into(),
+			}),
+		}
+	}
```

Comment:
AFAIK this will make `bytes::len("foobar")` no longer possible, which isn't the worst thing in the world, but just making sure it was intentional.

Reply:
We could go down two routes:
1. Make the `bytes::len()` function, and other `bytes::` functions accept a `Value` and then check in the function itself whether it's a `Strand` or a `Bytes` and act accordingly.
2. We could add an additional `impl FromArg for BytesOrStrand {` (create a new enum) implementation and try to coerce to `Bytes` or `Strand` in that implementation.

Reply:
While both of those would work, it's probably cleaner to just force users to use `bytes::len(<bytes> "foobar")`, which has the benefit of being more explicit.